### PR TITLE
Moves cleaning to an atom proc to afterattack signal

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
## About The Pull Request

* Cleaning off window blood is now handled by the Window itself
* Adds a pre_clean_callback to replace the should_clean arg (mostly), also adds a DO_NOT_CLEAN define that will prevent cleaning entirely, to remove the need to have checks in individual afterattacks and instead do it in the callback (primary example is mop, which shouldn't clean if it has no reagents or is attacking a bucket/janicart)
* Removes the atom cleaning proc, replaces it with a signal for afterattack on a cleaning tool, or bot's afterattack.

## Why It's Good For The Game

Atom procs like these are against what signals should be, hopefully this will be better.
I also noticed that rags check if the rag is in user to work, it seems like a rather useless check? Is it there for some Telekinesis thing or were rags meant to only clean things on the user (so the bartender can wipe down glasses)? I'll probably make that change in the future once I find out.

## Changelog

🆑 
code: Players are no longer told about washing off someone's mouth with soap, since any cleaning tool (including soap) already does it, even if they don't target the mouth.
/🆑

### CODE REVIEW

**Comment by kawoppi** [comment]
Did you check if cleaning something while you're already cleaning it doesn't remove the bubbles animation before the do_after finishes? The already_cleaning var was used to prevent that but I'm not sure if other changes have made it unnecessary.

**Comment by JohnFulpWillard** [comment]
> Did you check if cleaning something while you're already cleaning it doesn't remove the bubbles animation before the ongoing do_after finishes? The already_cleaning var was used to prevent that but I'm not sure if other changes have made it unnecessary.

I removed 1/2 times it was used so just made the other time it was used check the trait instead

**Comment by kawoppi** [comment]
> I removed 1/2 times it was used so just made the other time it was used check the trait instead

It's not so much about the var itself but the check before cutting the overlays at the end. I think the issue was that if you clean something and then clean it again before the first one finishes, the second cleaning skips the do_after and cuts the overlays before the first cleaning can finish it's do_after. That check made it so it would only cut the overlays if it had also added those overlays earlier.

**Comment by JohnFulpWillard** [comment]
> > I removed 1/2 times it was used so just made the other time it was used check the trait instead
> 
> It's not so much about the var itself but the check before cutting the overlays at the end. I think the issue was that if you clean something and then clean it again before the first one finishes, the second cleaning skips the do_after and cuts the overlays before the first cleaning can finish it's do_after. That check made it so it would only cut the overlays if it had also added those overlays earlier.

The code currently adds a trait but only removes it if they don't have the trait, I thought that was silly so I removed it, from my short testing it didn't seem like a problem but I'll test it more now that you brought it up.

**Comment by JohnFulpWillard** [comment]
Tested this stuff in-game and it works just as expected, however cleanbots were broken. I fixed those and also changed a few other stuff as well.

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 1]
```diff
@@ -267,7 +268,7 @@
 				shuffle = TRUE //Shuffle the list the next time we scan so we dont both go the same way.
 				path = list()
 			else
-				UnarmedAttack(target) //Rather than check at every step of the way, let's check before we do an action, so we can rescan before the other bot.
+				UnarmedAttack(target, TRUE) //Rather than check at every step of the way, let's check before we do an action, so we can rescan before the other bot.
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:271]: named arguments

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 2]
```diff
@@ -239,9 +240,9 @@
 	if(bot_cover_flags & BOT_COVER_EMAGGED) //Emag functions
 		var/mob/living/carbon/victim = locate(/mob/living/carbon) in loc
 		if(victim && victim == target)
-			UnarmedAttack(victim) // Acid spray
+			UnarmedAttack(victim, TRUE) // Acid spray
 		if(isopenturf(loc) && prob(15)) // Wets floors and spawns foam randomly
-			UnarmedAttack(src)
+			UnarmedAttack(src, TRUE)
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:245]: here too

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 3]
```diff
@@ -239,9 +240,9 @@
 	if(bot_cover_flags & BOT_COVER_EMAGGED) //Emag functions
 		var/mob/living/carbon/victim = locate(/mob/living/carbon) in loc
 		if(victim && victim == target)
-			UnarmedAttack(victim) // Acid spray
+			UnarmedAttack(victim, TRUE) // Acid spray
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:243]: here too

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 4]
```diff
@@ -121,7 +122,7 @@
 /mob/living/simple_animal/bot/cleanbot/examine(mob/user)
 	. = ..()
 	if(!weapon)
-		return
+		return .
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:125]: dont do that

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 5]
```diff
@@ -333,12 +334,10 @@
 /mob/living/simple_animal/bot/cleanbot/UnarmedAttack(atom/attack_target, proximity_flag, list/modifiers)
 	if(HAS_TRAIT(src, TRAIT_HANDS_BLOCKED))
 		return
-	. = ..()
 	if(ismopable(attack_target))
 		mode = BOT_CLEANING
 		update_icon_state()
-		var/turf/turf_to_clean = get_turf(attack_target)
-		start_cleaning(src, turf_to_clean, src)
+		..()
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:340]: did you mean to not set return value anymore

File: code/modules/reagents/reagent_containers/misc.dm [hunk 6]
```diff
@@ -131,29 +131,31 @@
...
 
-/obj/item/reagent_containers/cup/rag/afterattack(atom/A as obj|turf|area, mob/living/user,proximity)
-	. = ..()
-	if(!proximity)
+/obj/item/reagent_containers/cup/rag/afterattack(atom/target, mob/living/user, proximity_flag, click_parameters)
+	if(!proximity_flag)
 		return
-	if(iscarbon(A) && reagents?.total_volume)
-		var/mob/living/carbon/C = A
+	if(iscarbon(target) && reagents?.total_volume)
```
Comment (Fikou) [inline code/modules/reagents/reagent_containers/misc.dm:143]: early return here please

**Review by Fikou — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm (hunk 5 above)
Comment (JohnFulpWillard) [reply code/modules/mob/living/simple_animal/bot/cleanbot.dm:340]: It wasn't necessary to have

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm (hunk 4 above)
Comment (Fikou) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:125]: i hate mothblocks

**Pull request merged**